### ISSUE TITLE
Suggestion: user defined aggregations for df.groupby

### ISSUE THREAD

**Issue by chmp** [dask/dask#2340]
Any interest in user defined aggregations for `DataFrame.groupby(...)` and `Series.groupby(...)`? I'd be willing to implement it myself, but wanted gauge general interest before starting.

As an API, I envision offering the possibility of passing objects with attributes `chunk`, `aggregate`, `combine`, and `split_every`. Such objects could be given as arguments inside the aggregate spec. 
Their semantics would follow `DataFrame.reduction(..)`. For example:

```
first_value = dd.reduction(chunk=lambda s: s.iloc[0], aggregate=lambda s: s.iloc[0])

first_value_of_by_g = ddf.groupby(['g']).aggregate({'c': first_value})
```

The major issue I see here, that this would fix the current implementation of `aggregate` in terms of the steps chunk, aggregate, combine.

edit: come to think about it, the functions most definitely would need to be able to return dataframes to allow for keeping more than a single value of state, e.g., for mean.

**Comment by mrocklin** [comment]
The relevant internal API here is `GroupBy._aca_agg`.  It might be worth looking at how `GroupBy.sum` and friends are currently implemented.  Perhaps there is a way to reveal those functions nicely to users.

**Comment by chmp** [comment]
I had a look at the current `.aggregate` implementation - since I'm somewhat familiar :) - which also internally uses `aca`. Since I take it you are not generally opposed to the idea, I'll prototype an implementation and to have something to discuss on.

**Comment by mrocklin** [comment]
Great.  Thank you for taking this on.

**Comment by TomAugspurger** [comment]
Closed by #2344
